agent/internal/fluent/fluent.go

```diff
@@ -0,0 +1,502 @@
+	fluentLogs      []*aproto.RunMessage
+	fluentLogsCount int
+
+	wg      errgroupx.Group // Fluentbit-scoped group.
+	err     error
+	started chan struct{} // Closed when FluentBit is actually started.
+	Done    chan struct{} // Closed when FluentBit exits.
+}
+
+// Start constructs and runs a Fluent Bit daemon, and returns a handle to interact with it.
+func Start(
+	ctx context.Context,
+	opts options.AgentOptions,
+	mopts aproto.MasterSetAgentOptions,
+	dclient *docker.Client,
+) (*Fluent, error) {
+	f := &Fluent{
+		opts:  opts,
+		mopts: mopts,
+
+		log:    log.WithField("component", "fluentbit"),
+		docker: dclient,
+
+		fluentLogs: make([]*aproto.RunMessage, 50),
+		wg:         errgroupx.WithContext(context.Background()),
+		started:    make(chan struct{}),
+		Done:       make(chan struct{}),
+	}
+
+	f.wg.Go(func(ctx context.Context) error {
+		defer f.wg.Cancel()
+		switch err := f.run(ctx); {
+		case errors.Is(err, context.Canceled):
+			return nil
+		case err != nil:
+			return err
+		default:
+			return nil
+		}
+	})
+
+	go func() {
+		f.err = f.wg.Wait()
+		close(f.Done)
+	}()
+
+	select {
+	case <-f.started:
+		return f, nil
+	case <-f.Done:
+		return nil, f.err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// Error returns the error associated with an exit, if there is any. It returns nil when running.
+func (f *Fluent) Error() error {
+	return f.err
+}
+
+// Wait for the daemon to exit.
+func (f *Fluent) Wait() error {
+	return f.wg.Wait()
+}
+
+// Close the Fluent Bit daemon.
+func (f *Fluent) Close() error {
+	return f.wg.Close()
+}
+
+func (f *Fluent) run(ctx context.Context) error {
+	f.log.Trace("starting fluent container")
+	t0 := time.Now()
+	cID, err := f.startContainer(ctx)
+	if err != nil {
+		return fmt.Errorf("starting logging container: %w", err)
+	}
+	close(f.started)
+	log.Infof("Fluent Bit started in %s", time.Since(t0))
+
+	exitC, errC := f.docker.Inner().ContainerWait(
+		ctx,
+		cID,
+		dcontainer.WaitConditionNotRunning,
+	)
+
+	logs := make(chan *aproto.ContainerLog, 64)
+	loggroup := errgroupx.WithContext(ctx)
+	loggroup.Go(func(ctx context.Context) error {
+		defer close(logs)
+		return f.trackLogs(ctx, cID, logs)
+	})
+	loggroup.Go(func(ctx context.Context) error {
+		for {
+			select {
+			case log, ok := <-logs:
+				if !ok {
+					return nil
+				}
+				if log.RunMessage == nil {
+					continue
+				}
+				f.recordLog(log.RunMessage)
+				match := fluentLogLineRegexp.FindStringSubmatch(log.RunMessage.Value)
+				if match == nil {
+					continue
+				}
+				switch level, message := match[1], match[2]; level {
+				case "warn":
+					f.log.Warnf("Fluent Bit: %s", message)
+				case "error":
+					f.log.Errorf("Fluent Bit: %s", message)
+				}
+			case <-ctx.Done():
+				return fmt.Errorf("fluent bit canceled: %w", ctx.Err())
+			}
+		}
+	})
+	switch err := loggroup.Wait(); {
+	case errors.Is(err, context.Canceled):
+		f.log.WithError(err).Warn("orphaning fluent due to cancellation")
```

Comment:
why aren't we killing fluent container on cancellation?
isn't cancellation here the normal operation e.g. when the agent process is signaled SIGQUIT, i.e. normal exit?

Reply:
actually, you don't want to kill fluent on cancellation, just on failure you want to make sure you clean up. if you roll the agent to upgrade, you dont want logs to stop working for a short time.

Reply:
this logic is replicated on line 186